File: content.en/docs/notes/dataStuctures_Algorithms/sort.cpp
#include <iostream>

// This is not even an insertion sort
void unkonwn_sort(int arr[], int size){
    int output[size];

    // Initialize
    for (int i = 0; i < size; ++i) {
        output[i] = arr[i];
        arr[i] = 0;
    }

    // Loop through all numbers
    for(int i = 0; i < size; i++){
        int num = output[i];

        // For each number first shift it right
        for(int j = size-1; j > 0; j--){
            arr[j] = arr[j-1];
            // if the currently iterating number in the array is smaller, we add it
            // in front of it
            if(arr[j] != 0 && arr[j] < num){
                arr[j] = num;
                break;
            // if we did not find number bigger, we put it to the first position
            } else if (j == 1){
                arr[j-1] = num;
            }
        }
    }
}

void reverse_insertion_sort(int arr[], int size){
    for(int i = 1; i < size; i ++){
        int key = arr[i];
        int j = i - 1;
        while(j >= 0 && arr[j] < key){
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = key;
    }
}

// Time Complexity O(n^2)
void insertion_sort(int arr[], int size){
    for(int i = 1; i < size; i++){
        int key = arr[i];
        int j = i - 1;
        while(j >= 0 && arr[j] > key){
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = key;
    }
}

int main() {
    int arr[] = {6,2,7,4,1,10,3,8,5,9};
    int size = sizeof(arr) / sizeof(arr[0]);

    std::cout << "Before sorting: ";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;

    //unkonwn_sort(arr, size);
    //insertion_sort(arr, size);
    reverse_insertion_sort(arr, size);

    std::cout << "After sorting: ";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}

/* Testing:
int test(){
    int arr[] = {1, 2, 3, 4, 5, 6};
    int size = sizeof(arr) / sizeof(arr[0]); --> in this case size = 6

    // Returns: total byte size: (e.g: 6 elements x 4 bytes = 24)
    std::cout << sizeof(arr) << std::endl;

    // Returns: memory addresss (e.g: 0x5ffe80)
    std::cout << arr << std::endl;

}
*/

File: content.en/docs/notes/dataStuctures_Algorithms/sort.py
import random
import time

# Concept of loop invariant: conditions that holds true before, during, and after every iteration of a loop.



# Time Complexity O(n^2) --> This method is still better than the "unknown sort"
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while(j >= 0 and arr[j] > key):
            arr[j+1] = arr[j]
            j-=1
        arr[j+1] = key
    return arr

# Time Complexity O(n^2) --> 
def unknown_sort(array):
    output = [0] * len(array)
    
    for i in range(len(array)):
        num = array[i]
        # case 1: it is the start
        if(i == 0):
            output[i] = num
        # case 2: it is not the start
        if(i > 0):
            for j in range(len(array)-1, 0, -1):
                if(output[j-1] != 0):
                    output[j] = output[j-1]
                    if(output[j] < num):
                        output[j] = num
                        break
                    if(j == 1):
                        output[j-1] = num
                        break
    return output


array = [random.randint(1, 10) for _ in range(10000)]

start = time.time()
sorted_unknown = unknown_sort(array)
end = time.time()
print("Time for unknown sort: ", (end-start)*10**3, "ms")

start = time.time()
sorted_insertion = insertion_sort(array)
end = time.time()
print("Time for insertion sort: ", (end-start)*10**3, "ms")

#print(sorted_unknown)
#print(sorted_insertion)
    